src/ConstantOptimization.jl

```diff
@@ -20,16 +20,19 @@ end
 
 # Use Nelder-Mead to optimize the constants in an equation
 function optimize_constants(
-    dataset::Dataset{T}, member::PopMember{T}, options::Options
-)::Tuple{PopMember{T},Float64} where {T<:Real}
+    dataset::Dataset{T,L}, member::PopMember{T,L}, options::Options
+)::Tuple{PopMember{T,L},Float64} where {T<:DATA_TYPE,L<:LOSS_TYPE}
     nconst = count_constants(member.tree)
     num_evals = 0.0
     if nconst == 0
         return (member, 0.0)
     end
     x0 = get_constants(member.tree)
-    f(x::Vector{T})::T = opt_func(x, dataset, member.tree, options)
-    if nconst == 1
+    f(x::Vector{T})::L = opt_func(x, dataset, member.tree, options)
+    if T <: Complex
+        # TODO: Make this more general. Also, do we even need Newton here at all??
```

Comment:
I'm not sure why the other would be Newton. Is there a reason for preferring Newton in the Real case? With finite differences for the Hessian, I'd suspect it's not actually doing much better. If anything, doing a single Hessian to pre-start the BFGS could get similar robustness for less work.

Reply:
Thanks - If I remember correctly, I found Newton was cheaper for 1-parameter optimizations than the others, so seemed like a good default for simple equations. (It will be run multiple times throughout the search). But I don't remember specifics, I should really measure this again.

For complex numbers, of these 3, it seems like only BFGS is implemented:

```julia
using Optim

f(x) = (1.0 - x[1])^2 + 100.0 * (x[2] - x[1]^2 + 1im)^2

x0 = zeros(ComplexF64, 2)

optimize(x -> abs2(f(x)), x0, Newton())
# MethodError: no method matching eps(::Type{ComplexF64})
optimize(x -> abs2(f(x)), x0, NelderMead())
# MethodError: no method matching isless(::ComplexF64, ::ComplexF64)

optimize(x -> abs2(f(x)), x0, BFGS())
# Works!
```

---

(Btw, another to-do I have is to allow the user to pass an arbitrary optimizer.)

Reply:
ahh yeah for one parameter optimizations or very small then Newton can be fine. With complex, the Hessian is a bit hard to define because it's dependent on if the function is analytic or not (in the non-analytic case it needs to have a much larger system), so I don't think that will be solved any time soon. So BFGS is going to be needed with the complex case.